### PR TITLE
Default affinity

### PR DESCRIPTION
**Pull request by jhamman**
At this point, this is for illustration purposes only. It builds on #145 and gives #146 a try. By default, this now creates dask-worker pods with the following affinity and tolerations:

```yaml
spec:
  affinity:
    nodeAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - preference:
          matchExpressions:
          - key: k8s.dask.org/node-purpose
            operator: In
            values:
            - worker
        weight: 100
  tolerations:
  - effect: NoSchedule
    key: k8s.dask.org/dedicated
    operator: Equal
    value: worker
  - effect: NoSchedule
    key: k8s.dask.org_dedicated
    operator: Equal
    value: worker
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
```

If my understanding is correct: A kubernetes cluster / node-pool with the following labels/taints should attract and accept dask-workers with this spec.

labels:
 - `k8s.dask.org/node-purpose : worker`

taints:
- `k8s.dask.org/dedicated=worker`
- `k8s.dask.org_dedicated=worker`

I should note that I have a node-pool with these options set and my dask-workers are not landing in the desired place yet so I'm probably still missing at least one piece of the puzzle.

### CODE REVIEW

**Comment by mrocklin** [comment]
@quasiben if you have a moment can I ask you to check out this PR?

**Comment by jhamman** [comment]
Thanks @jacobtomlinson - I need to add a unit test or two but now that its passed your sniff test, I can forge ahead.

**Comment by jhamman** [comment]
I'm working on tests now. I've also created a binder with this branch on it that can be run on binder.pangeo.io: [Binder](https://binder.pangeo.io/v2/gh/jhamman/pangeo_test_binder/dask_k8s_affinity)

File: dask_kubernetes/objects.py [hunk 1]
```diff
@@ -222,4 +206,49 @@ def clean_pod_template(pod_template):
...
+    )
+    
+    if pod_template.spec.affinity is None:
+        pod_template.spec.affinity = client.V1Affinity(
+                node_affinity=client.V1PodAntiAffinity(
+                    preferred_during_scheduling_ignored_during_execution=[affinity]
+                )
+            )
+    else:
+        print('TODO: need to handle the various partial states here')
```
Comment (consideRatio) [inline dask_kubernetes/objects.py]: I've done partial state handling in #148, doing this was really messy sadly.

Comment (consideRatio) [reply dask_kubernetes/objects.py]: You are very welcome to copy paste etc to accomplish the desired outcome.

**Comment by jhamman** [comment]
@consideRatio - I merged our two efforts. Let me know what you think.

File: dask_kubernetes/objects.py [hunk 2]
```diff
@@ -229,4 +229,56 @@ def clean_pod_template(pod_template):
...
+        if affinity is None:
+            affinity = client.V1Affinity()
+        if affinity.node_affinity is None:
+            affinity.node_affinity = client.V1NodeAffinity()
+
+        # a common object for both a preferred and a required node affinity
+        node_selector_term = client.V1NodeSelectorTerm(
+            match_expressions=[
+                client.V1NodeSelectorRequirement(
+                    key="k8s.dask.org/node-purpose", operator="In", values=["worker"]
```
Comment (jhamman) [inline dask_kubernetes/objects.py]: are we okay with In vs. Equals here?

Comment (consideRatio) [reply dask_kubernetes/objects.py]: Logically right now they are exactly the same, I think it is possible to make it `"Equal"` (or was it `"Equals"`?) instead assuming we remove the brackets `[]` around `"worker"`.

File: dask_kubernetes/objects.py [hunk 3]
```diff
@@ -229,26 +229,56 @@ def clean_pod_template(pod_template):
...
-    else:
-        print('TODO: need to handle the various partial states here')
+        elif match_node_purpose == "prefer":
+            if (
+                affinity.node_affinity.preferred_during_scheduling_ignored_during_execution
+                is None
+            ):
+                affinity.node_affinity.preferred_during_scheduling_ignored_during_execution = (
+                    []
+                )
```
Comment (consideRatio) [inline dask_kubernetes/objects.py]: The assignment above should be an array I think. If we could say the type of elements it should be an array of V1PreferredSChedulingTerm objects, but as we cannot type the arrays element it should simply be an array.

I understand it currently as it is a tuple with one element: an array. I'm not confident on how the parenthesis are interpreted though.

File: dask_kubernetes/objects.py (hunk 2 above)
Comment (jhamman) [reply dask_kubernetes/objects.py]: Looking at this a little more, it doesn't seem like Equals is allowed here: https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1NodeSelectorRequirement.md

Comment (consideRatio) [reply dask_kubernetes/objects.py]: Ah

File: dask_kubernetes/objects.py (hunk 3 above)
Comment (jhamman) [reply dask_kubernetes/objects.py]: This evaluates to an empty list.

```ipython
In [1]: x = ([])

In [2]: x
Out[2]: []
```

As you say, we can't type this further. AFAIK, there isn't a way to instantiate an empty list of `V1PreferredSchedulingTerm` objects, like we do for the `V1NodeSelector`s above.

Comment (consideRatio) [reply dask_kubernetes/objects.py]: Ah okay! Technically then all good!!

I just learned that `()` is a 0-tuple or a tuple with 0 elements, and ("test",) is a tuple with 1 element, and ("test", "more") is a tuple with two elements, but ("something") is just "something".

**Comment by jhamman** [comment]
@yuvipanda / @jacobtomlinson - I think we're all done here. Ready to merge?

**Pull request merged**